### DISCUSSION TITLE
How to make use of the NuGet uplink for package dependencies?

### DISCUSSION THREAD

**Discussion by cmrgKoradir** [openupm/openupm#2414] [Q&A] [answered]
We're using [Sentry](https://github.com/getsentry/unity). Sentry comes with some bundled dependencies like e.g. the `System.Threading.Tasks.Extensions`, which - surprise - conflicts with a different package that also includes the dependency (albeit apparently on a different version).

I understand that's one of the motivations behind https://github.com/openupm/openupm/issues/1976

What I don't understand yet is how to make use of it.

So I [forked sentry](https://github.com/cmrgKoradir/cmrg-sentry-unity) with the intention of removing the .dll files and replacing them with a dependency on the NuGet packages.

I thought that would be simple as I'd just need to delete the .dll files in question, and toss some `dependencies` into the `package.json` file (as I did [here](https://github.com/cmrgKoradir/cmrg-sentry-unity/commit/78c9292d021a2c88e41a8511769925978ea845b8)).

I'm still trying to debug this whole thing, so I intend to install it using the git url.
So after failing a couple of times, I install the UPM Extension for Git, just to be sure it's there if it should be required.

```
> openupm install git@github.com:cmrgKoradir/cmrg-sentry-unity.git
notice manifest added git@github.com:cmrgKoradir/cmrg-sentry-unity.git
notice please open Unity project to apply changes
```

Alright, so far so good, except ... that's not actually valid according to Unity when I try to start it up

> An error occurred while resolving packages:
  Project has invalid dependencies:
    git: Version 'github.com:cmrgKoradir/cmrg-sentry-unity.git' is invalid. Expected one of: a 'SemVer' compatible value; a value starting with 'file:'; a Git URL starting with 'git:' or 'git+', or ending with '.git'.

fine, so let's change that to 
```
"git": "git@github.com:cmrgKoradir/cmrg-sentry-unity.git"
```
and then to
```
"io.sentry.unity": "git@github.com:cmrgKoradir/cmrg-sentry-unity.git"
```
because 
> The requested dependency 'git' does not match the `name` 'io.sentry.unity' specified in the package manifest of [git@git@github.com:cmrgKoradir/cmrg-sentry-unity.git]

(the reason I'm mentioning all this is because I have no idea if any of this has any effect on the package resolution)

Finally, let's start up unity again aaaaand....

```
Package Manager log level set to [2]
[Package Manager] Done registering packages in 0.10s seconds
[Package Manager] Done resolving packages in 7.87s seconds
An error occurred while resolving packages:
  Package io.sentry.unity@git@github.com:cmrgKoradir/cmrg-sentry-unity.git has invalid dependencies or related test packages:
    org.nuget.system.buffers (dependency): Package [org.nuget.system.buffers@4.5.1] cannot be found
    org.nuget.system.collections.immutable (dependency): Package [org.nuget.system.collections.immutable@5.0.0] cannot be found
    org.nuget.system.memory (dependency): Package [org.nuget.system.memory@4.5.4] cannot be found
    org.nuget.system.numerics.vectors (dependency): Package [org.nuget.system.numerics.vectors@4.5.0] cannot be found
    org.nuget.system.reflection.metadata (dependency): Package [org.nuget.system.reflection.metadata@5.0.0] cannot be found
    org.nuget.system.runtime.compilerservices.unsafe (dependency): Package [org.nuget.system.runtime.compilerservices.unsafe@5.0.0] cannot be found
    org.nuget.system.text.encodings.web (dependency): Package [org.nuget.system.text.encodings.web@5.0.1] cannot be found
    org.nuget.system.text.json (dependency): Package [org.nuget.system.text.json@5.0.2] cannot be found
    org.nuget.system.threading.tasks.extensions (dependency): Package [org.nuget.system.threading.tasks.extensions@4.5.4] cannot be found

A re-import of the project may be required to fix the issue or a manual modification of /Users/koradir/code/neo-medical-app/NeoMedicalUnity/Packages/manifest.json file.
```

Now openupm does find the dependencies when I `openupm view` them, but ... how do I get package resolution to work, here?

**Comment by favoyang** [comment, marked as answer]
As mentioned in openupm-cli's README
> The add command doesn't verify package or resolve dependencies for Git, HTTPS, and file protocol.

When using a Gi URL ([learn more here](https://docs.unity3d.com/Manual/upm-git.html)), the openupm-cli simply added it to the manifest.json and won't resolve the dependencies. So Unity Editor can not find the org.nuget scopes.

However, when installing a package available on OpenUPM, the openupm-cli will do the right thing - all the dependent org.nuget.* packages are added to the scopes list.

```
$ openupm add com.example.nuget-consumer
...

$ cat Packages/manifest.json
 "scopedRegistries": [                                            
   {                                                              
     "name": "package.openupm.com",                               
     "url": "https://package.openupm.com",                        
     "scopes": [                                                  
       "com.example.nuget-consumer",                              
       "com.openupm",                                             
       "org.nuget.system.buffers",                                
       "org.nuget.system.memory",                                 
       "org.nuget.system.numerics.vectors",                       
       "org.nuget.system.runtime.compilerservices.unsafe"         
     ]                                                            
   }                                                              
```

So for testing purposes against Git, you can manually add a scope entry "org.nuget".

```
 "scopedRegistries": [                                            
   {                                                              
     "name": "package.openupm.com",                               
     "url": "https://package.openupm.com",                        
     "scopes": [                                                                    
       "com.openupm",
       "org.nuget"
     ]                                                            
   }                
```

BTW, https://github.com/mob-sakai/UpmGitExtension is a dedicated Git solution, perhaps can do some magic resolving for Git URLs.

**Comment by cmrgKoradir** [comment]
Adding the scope entry did the trick, thank you.
